internal/querynodev2/tasks/queues.go

```diff
@@ -0,0 +1,209 @@
+package tasks
+
+import (
+	"container/ring"
+	"time"
+)
+
+func newMergeTaskQueue(group string) *mergeTaskQueue {
+	return &mergeTaskQueue{
+		name:             group,
+		tasks:            make([]Task, 0),
+		cleanupTimestamp: time.Now(),
+	}
+}
+
+type mergeTaskQueue struct {
+	name             string
+	tasks            []Task
+	cleanupTimestamp time.Time
+}
+
+// len returns the length of taskQueue.
+func (q *mergeTaskQueue) len() int {
+	return len(q.tasks)
+}
+
+// push add a new task to the end of taskQueue.
+func (q *mergeTaskQueue) push(t Task) {
+	q.tasks = append(q.tasks, t)
+}
+
+// front returns the first element of taskQueue,
+// returns nil if task queue is empty.
+func (q *mergeTaskQueue) front() Task {
+	if q.len() > 0 {
+		return q.tasks[0]
+	}
+	return nil
+}
+
+// pop pops the first element of taskQueue,
+func (q *mergeTaskQueue) pop() {
+	if q.len() > 0 {
+		q.tasks = q.tasks[1:]
+		if q.len() == 0 {
+			q.cleanupTimestamp = time.Now()
+		}
+	}
+}
+
+// Return true if user based task is empty and empty for d time.
+func (q *mergeTaskQueue) expire(d time.Duration) bool {
+	if q.len() != 0 {
+		return false
+	}
+	if time.Since(q.cleanupTimestamp) > d {
+		return true
+	}
+	return false
+}
+
+// tryMerge try to a new task to any task in queue.
+func (q *mergeTaskQueue) tryMerge(task MergeTask, maxNQ int64) bool {
+	nqRest := maxNQ - task.NQ()
+	// No need to perform any merge if task.nq is greater than maxNQ.
+	if nqRest < 0 {
+		return false
+	}
+	for i := 0; i < q.len(); i++ {
```

Comment:
reverse iteration takes higher probability to return early

Reply:
- Reverse iteration do not make great improvement on my last test.
- Task of same segment may locate at random positions in the queue.